addons/addon-raas-s3-copy/packages/s3-synchronizer/src/s3-synchronizer_test.go

```diff
@@ -142,14 +148,273 @@ func TestMainImplForInitialDownloadEmptyMounts(t *testing.T) {
 // Negative test: Test for invalid s3Mounts json
 func TestMainImplForInitialDownloadInvalidMounts(t *testing.T) {
 	// ---- Inputs ----
-	debug := true
 	concurrency := 2
 
 	testMountsJson := "some invalid json"
 	fmt.Printf("Input: \n\n%s\n\n", testMountsJson)
 
 	// ---- Run code under test ----
-	err := mainImpl(testAwsSession, debug, concurrency, testMountsJson, destinationBase)
+	err := mainImpl(testAwsSession, debug, false, -1, 60, concurrency, testMountsJson, destinationBase)
+	if err == nil {
+		// Fail test in case of no errors since we are expecting errors when passing invalid json for mounting
+		t.Logf("Expecting error when running the main s3-synchronizer with invalid testMountsJson but it ran fine")
+	}
+}
+
+// ######### Tests for Recurring Downloads #########
```

Comment:
Can we add one test for error handling in recurring downloads. Say S3 had some transient issue and sync method fails the first time but the expectation is that a re-sync would happen after downloadInterval is passed and hopefully succeed.

Reply:
Good suggestion. Would be certainly be a nice test to have. Unfortunately, I couldn't find any straight forward way to simulate intermittent S3 failures. The test is using [gofakes3](https://github.com/johannesboyne/gofakes3) for simulating AWS S3 APIs. Between mocking S3 SDK and simulating S3 APIs, I had chosen "simulating" due to it being non intrusive and its ease of implementation.

For simulating intermittent S3 failures, here is what I tried without luck
1. Stop (Close) and Start the gofakes3 server  -- Result: gofakes3 server cannot be restarted.
2. Close and completely replace gofakes3 server with a new server -- Result: new gofakes3 server starts on another port but S3 session is already pinned to previous endpoint port.
3. Change the Endpoint in the AWS SDK Session to point to some non-existent URL and then swap it back to the gofakes3 url: The Endpoint swap never materializes after first usage. AWS Session caches it. 

Due to these, I will not invest more time on adding this test at the moment. I have tested S3 failure manually (not intermittent though), I have manually verified that the synchronizer thread does not exit and it backs-off for about 30 seconds and then attempts to synchronize again.

Reply:
Thanks for trying all the various approaches. I think the only way to make this work is probably to inject a new S3 session every time we synchronize. Is that feasible? On a related note re-using s3 session is credential rotation safe, right?

Reply:
Creating new AWS Session every time is an anti-pattern. AWS SDKs in almost all languages have some concept of "initialization" where you "initialize" the SDK with some configuration once and then keep creating service clients and interacting with those clients. 

One way to make this work would be to refactor the program especially the s3-downloader and s3-synchronizer to hold the session in a state (e.g., some struct) and then make the functions read the session from that state instead of passing the session as function argument. We can then swap out the session in that state in the test code temporarily to simulate failures and then swap it back. Since that would require some non-trivial refactoring. I will add a ticket for that.

Reply:
Sounds good to me, you can resolve this conversation once the ticket is created.

Reply:
Done. Created a ticket.